sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSourceStrategy.scala

```diff
@@ -256,20 +256,32 @@ class FindDataSourceTable(sparkSession: SparkSession) extends Rule[LogicalPlan]
       QualifiedTableName(table.identifier.catalog.get, table.database, table.identifier.table)
     val catalog = sparkSession.sessionState.catalog
     val dsOptions = DataSourceUtils.generateDatasourceOptions(extraOptions, table)
-    catalog.getCachedPlan(qualifiedTableName, () => {
+    lazy val newPlan = {
       val dataSource =
         DataSource(
           sparkSession,
-          // In older version(prior to 2.1) of Spark, the table schema can be empty and should be
-          // inferred at runtime. We should still support it.
-          userSpecifiedSchema = if (table.schema.isEmpty) None else Some(table.schema),
+          className = table.provider.get,
+          userSpecifiedSchema = Some(table.schema),
           partitionColumns = table.partitionColumnNames,
           bucketSpec = table.bucketSpec,
-          className = table.provider.get,
           options = dsOptions,
           catalogTable = Some(table))
       LogicalRelation(dataSource.resolveRelation(checkFilesExist = false), table)
-    })
+    }
+    lazy val cachedPlan = catalog.getCachedPlan(qualifiedTableName, () => newPlan)
```

Comment:
Nit: can skip lazy val here and just use the expression to match directly?

catalog.getCachedPlan ... match {

Reply:
Sorry , clarifying one possible way (from discussion offline) 

catalog.getCachedPlan(...) match {
   case LogicalRelation(HadoopFsRelation(... options)) if diff(options, dsOptions) => newPlan
   case cachedPlan => cachedPlan
 }
 
Its slightly clearer to me, but optional. 

If we use a val though, I think we dont need lazy val as it wasnt lazy val before this change?